### PR TITLE
Rfs instrumentation

### PR DESCRIPTION
**Pull request by gregschohn**
### Description
Instruments RestClient within RFS so that every call has a context.  Instrument work coordination so that every low-level and high-level interaction has a scope and metrics are emitted appropriately.

* Category New feature/Refactoring/OE
* Why these changes are required? To get better observability into RFS
* What is the old behavior before changes and new behavior after changes? Now HTTP calls in RFS emit spans and metrics.

*Still TODO* - verify that the metrics for the work coordinator are as expected.

### Issues Resolved
https://opensearch.atlassian.net/browse/MIGRATIONS-1738

Is this a backport? If so, please add backport PR # and/or commits #

### Testing
Limited manual testing.  RFS doesn't have much test infrastructure yet

### Check List
- [ ] New functionality includes testing
  - [ ] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: RFS/src/main/java/com/rfs/tracing/RootRfsContext.java
Comment (chelma) [inline RFS/src/main/java/com/rfs/tracing/RootRfsContext.java:12]: Why two of this type?

**Review by chelma — COMMENTED** [review]
Spent some time looking things over and left some comments, but probably better to discuss verbally.  Had trouble piecing together the bigger picture here.

File: RFS/src/main/java/com/rfs/tracing/RootRfsContext.java
Comment (gregschohn) [reply RFS/src/main/java/com/rfs/tracing/RootRfsContext.java:12]: fixed in a later commit

File: RFS/src/main/java/com/rfs/tracing/RfsContexts.java
Comment (gregschohn) [reply RFS/src/main/java/com/rfs/tracing/RfsContexts.java:208]: It's what you asked me to instrument in task MIGRATIONS-1738

File: RFS/src/main/java/com/rfs/tracing/RfsContexts.java
Comment (gregschohn) [reply RFS/src/main/java/com/rfs/tracing/RfsContexts.java:191]: I did a `s/IndexTemplateContext/TemplateContext/` in a later commit

File: RFS/src/main/java/com/rfs/tracing/IRfsContexts.java
Comment (gregschohn) [reply RFS/src/main/java/com/rfs/tracing/IRfsContexts.java:56]: Different how?  Those calls do a get, then a put, so there are potentially multiple requests, the IChecked...Context can create request scopes as needed for those two kinds of calls.

File: RFS/src/main/java/com/rfs/tracing/RfsContexts.java
Comment (gregschohn) [reply RFS/src/main/java/com/rfs/tracing/RfsContexts.java:206]: I'm not sure what your enumeration is... I'll admit that having methods to create a context that is in reference to an HTTP request that is creating something is confusing.  Given that this code is already changing underfoot and that the task was to establish the instrumentation pattern, this is something that can be picked up later, right?

File: RFS/src/main/java/com/rfs/RunRfsWorker.java
Comment (peternied) [inline RFS/src/main/java/com/rfs/RunRfsWorker.java:102]: Blocking; I know how to visualize the output of these instrumentation, can you update the readme or update the code to allow for a very basic log things to files?  I am OK if the dump isn't a file but a running Jager instance as long as its easy for me to spin up and connect.

File: RFS/src/test/java/com/rfs/tracing/TestContext.java
Comment (peternied) [inline RFS/src/test/java/com/rfs/tracing/TestContext.java:8]: There are a couple of unused methods here - lets either remove them or lets refactor to converge on a single TestContext factory.

File: RFS/src/test/java/com/rfs/integration/SnapshotStateTest.java
Comment (peternied) [inline RFS/src/test/java/com/rfs/integration/SnapshotStateTest.java:64]: Nit; this should be refactored into the SimpleRestoreFromSnapshot_ES_7_10.  I'll do this later - thanks for merging this in.

File: RFS/src/test/java/com/rfs/common/RestClientTest.java
Comment (peternied) [inline RFS/src/test/java/com/rfs/common/RestClientTest.java:31]: Creating the scope in this case doesn't seem as useful since you only operating inside a single scope.  If we wanted to 'prove' that the scopes were propigated, lets add a second scope where postAsync is used that could be differentiate on, what do you think?

File: RFS/src/main/java/com/rfs/RunRfsWorker.java
Comment (peternied) [inline RFS/src/main/java/com/rfs/RunRfsWorker.java:103]: Looks like we are using kabob casing

```suggestion
                names = {"--otel-collector-endpoint"},
```

File: RFS/src/main/java/com/rfs/version_os_2_11/GlobalMetadataCreator_OS_2_11.java
Comment (peternied) [inline RFS/src/main/java/com/rfs/version_os_2_11/GlobalMetadataCreator_OS_2_11.java:103]: What do you think of rewriting with try with resource pattern rather than a commment? E.g:

```java
for (String templateName : templateWhitelist) {
    // ...
    try (final var context = context.createComponentTemplateContext()) {
        client.createComponentTemplate(templateName, settings, context);
    } 
}
```

Comment (peternied) [reply RFS/src/main/java/com/rfs/version_os_2_11/GlobalMetadataCreator_OS_2_11.java:103]: After revisiting I would recommend an alternative to make the method `createComponentTemplate(...)`  responsiblity to create the context

File: RFS/src/main/java/com/rfs/tracing/RootRfsContext.java
Comment (peternied) [inline RFS/src/main/java/com/rfs/tracing/RootRfsContext.java:16]: 1) Why are all of these made and accessible outside the root context, I'm not sure what the case for access is?
2) Shouldn't they be created and passed into `create***Context(...)`?  Note if this is about singular creation, lets use a lazy supplier

File: RFS/src/main/java/com/rfs/tracing/RfsContexts.java
Comment (peternied) [inline RFS/src/main/java/com/rfs/tracing/RfsContexts.java:385]: As a project go all in or out on these annotations.  Want to raise an issue/ticket on this?

File: RFS/src/main/java/com/rfs/tracing/RfsContexts.java
Comment (peternied) [inline RFS/src/main/java/com/rfs/tracing/RfsContexts.java:30]: What is a call type, what does it represent?

File: RFS/src/main/java/com/rfs/tracing/RfsContexts.java
Comment (peternied) [inline RFS/src/main/java/com/rfs/tracing/RfsContexts.java:171]: Maybe this is for flexiblity, but the method that calls `createRegisterRequest()` is registerSnapshotRepo, seems like the name should be from the calling method annotated onto the GenericRequestContext.

What I would consider a better alternative is to have the caller provide the label, e.g.
```
createRegisterRequest(...) {
     ...
     
     try (final var requestContext = context.createRequestContext("createRegisterRequest")) {
        // Use it
     }
     ...
}
```

It would be even better if this was automatically using the callstack

File: RFS/src/main/java/com/rfs/tracing/RfsContexts.java
Comment (peternied) [reply RFS/src/main/java/com/rfs/tracing/RfsContexts.java]: @gregschohn this is something I called out in another comment.  If the the code changes, should the context name change with it, or do we need to be intentional about the changes to telemetry?

Note; I think the answer is it depends.
- If there is a specific contract to uphold, consider if there is an external system that expects `RFS.DocumentProcessor.DocsRead.Count` we should not break that system even if we rename DocumentProcessor to LuceneDocumentsProcessor
- If there is no contract we should limit the amount of 'legacy' that can build up in our systems.

Comment (peternied) [reply RFS/src/main/java/com/rfs/tracing/RfsContexts.java]: For the current part of the lifecycle I would rather we have less flexiblity and less to cleanup as we quickly iterate on this first productized version, we could always come back later.

File: RFS/src/main/java/com/rfs/tracing/RfsContexts.java
Comment (peternied) [reply RFS/src/main/java/com/rfs/tracing/RfsContexts.java]: @chelma @gregschohn do we need to revisit how/what these context mean?

**Review by peternied — CHANGES_REQUESTED** [review]

File: RFS/src/main/java/com/rfs/RunRfsWorker.java
Comment (peternied) [reply RFS/src/main/java/com/rfs/RunRfsWorker.java:102]: Resolved by this PR https://github.com/gregschohn/opensearch-migrations/pull/4

Comment (peternied) [reply RFS/src/main/java/com/rfs/RunRfsWorker.java:102]: Since I was able to create that PR I do not consider this to be a blocking issue anymore - resolving this thread.

**Review by peternied — APPROVED** [review]
@gregschohn Thanks for the conversation yesterday about the outstanding comments, I'll leave it to you if you'd prefer to merge or address any of the comment threads.

I created the following PRs against this branch on your fork, feel free to merge or close - but I'd value your thoughts

- https://github.com/gregschohn/opensearch-migrations/pull/3
- https://github.com/gregschohn/opensearch-migrations/pull/4
- https://github.com/gregschohn/opensearch-migrations/pull/5

File: RFS/src/main/java/com/rfs/tracing/BaseRootRfsContext.java
Comment (peternied) [inline RFS/src/main/java/com/rfs/tracing/BaseRootRfsContext.java:7]: We've got many stand alone apps that are all under the migration space, what about MigrationAppContext? If we are following the model that everything should be shared, shouldn't this migrate out into the core utils space?

File: MetadataMigration/src/main/java/com/rfs/MetadataMigration.java
Comment (AndreKurait) [inline MetadataMigration/src/main/java/com/rfs/MetadataMigration.java:91]: do you mean scheme:host:port?

Comment (gregschohn) [reply MetadataMigration/src/main/java/com/rfs/MetadataMigration.java:91]: We pass the argument onto [this call](https://javadoc.io/doc/io.opentelemetry/opentelemetry-exporter-otlp/0.12.0/io/opentelemetry/exporter/otlp/OtlpGrpcSpanExporter.Builder.html#setEndpoint(java.lang.String)) in Otel.  It simply says "Sets the OTLP endpoint to connect to. Optional, defaults to "localhost:4317".

Since the endpoint runs OTLP, specifying http or https is probably incorrect.  We should remove them from our docs and our deployment code (& double check).

File: CreateSnapshot/src/main/java/com/rfs/CreateSnapshot.java
Comment (peternied) [reply CreateSnapshot/src/main/java/com/rfs/CreateSnapshot.java:75]: Also lets use a standard format for arguments, in this case `--otel-collector-endpoint`

File: DocumentsFromSnapshotMigration/src/main/java/com/rfs/RfsMigrateDocuments.java
Comment (peternied) [inline DocumentsFromSnapshotMigration/src/main/java/com/rfs/RfsMigrateDocuments.java:163]: There are multiple `root` contexts at the same time?  Lets drop the word root unless it means something, `WorkCoordinationContext` and `DocumentMigrationContext`

File: RFS/src/main/java/org/opensearch/migrations/snapshot/creation/tracing/CreateSnapshotContext.java
Comment (peternied) [inline RFS/src/main/java/org/opensearch/migrations/snapshot/creation/tracing/CreateSnapshotContext.java:1]: This extra layer `snapshot.creation` doesn't make much sense to me.  If this package names are weird, its because we are working backward from functionality in an executable in a shard library space rather treating this as a shared library with a clear sense of what context is required at each level.

I'd rather have generic IContext use in the library and concrete implementation in the executable project so we can keep this clearer.  What do you think?

File: DocumentsFromSnapshotMigration/src/main/java/com/rfs/RfsMigrateDocuments.java
Comment (peternied) [inline DocumentsFromSnapshotMigration/src/main/java/com/rfs/RfsMigrateDocuments.java:222]: This is hacky, we've got a instance that should accept a context and internally it can create the work item pending context.

File: DocumentsFromSnapshotMigration/src/main/java/com/rfs/RfsMigrateDocuments.java
Comment (peternied) [inline DocumentsFromSnapshotMigration/src/main/java/com/rfs/RfsMigrateDocuments.java:235]: The documents runner to have a context passed into it, no?

Also the way `DocumentsRunner` is written is hard to understand, lets reduce the inlining

**Review by peternied — COMMENTED** [review]
Quick hit of comments, still reviewing

File: DocumentsFromSnapshotMigration/src/main/java/com/rfs/RfsMigrateDocuments.java
Comment (gregschohn) [reply DocumentsFromSnapshotMigration/src/main/java/com/rfs/RfsMigrateDocuments.java:243]: Why is it hacky?  The lifecycle of the context should be controlled by the work coordinator.  It seems more hacky to make every call to the work coordinator be responsible for creating and tearing down contexts around singly-scoped calls.

File: RFS/src/main/java/org/opensearch/migrations/snapshot/creation/tracing/CreateSnapshotContext.java
Comment (gregschohn) [reply RFS/src/main/java/org/opensearch/migrations/snapshot/creation/tracing/CreateSnapshotContext.java:1]: I plan on moving everything in this directory to the executable project.  I'm doing the class refactorings now, but holding off on the file moves because additional files will need to move around rather than just the ones that I've created here.  I can do that as a fast-follow, but I'd like to do it in such a way to minimize conflicts.

As per the package name - "snapshot.creation" didn't sit completely at ease with me.  I wasn't sure what this package should be called longer term (it definitely shouldn't be 'com.rfs', nor should it likely go under the top-level "org.opensearch.migrations" - I'm open to other suggestions.

File: DocumentsFromSnapshotMigration/src/main/java/com/rfs/RfsMigrateDocuments.java
Comment (gregschohn) [reply DocumentsFromSnapshotMigration/src/main/java/com/rfs/RfsMigrateDocuments.java:178]: Root does have meaning - the root context is a constructor and the root of the scope-trees for a given component.  Every scope needs a root, directly or indirectly.  From within the work coordination code (which should be considered its own independent component), there needs to be a root, hence the root coordination context.  

However, the doc migration code has its own containing context, hence the reason to splice the root and scopes from work coordination into the reindexing contexts when appropriate.

File: CreateSnapshot/src/main/java/com/rfs/CreateSnapshot.java
Comment (gregschohn) [reply CreateSnapshot/src/main/java/com/rfs/CreateSnapshot.java:75]: Ugh - I've changed this for now, but we have to switch everything over to be camel case.  This will be a bug waiting to happen.  See https://opensearch.atlassian.net/browse/MIGRATIONS-1837

File: DocumentsFromSnapshotMigration/src/main/java/com/rfs/RfsMigrateDocuments.java
Comment (peternied) [reply DocumentsFromSnapshotMigration/src/main/java/com/rfs/RfsMigrateDocuments.java:243]: Its hacky because it requires the caller RfsMigrateDocuments needs to know how to generate the correct kind of contexts for the work coordinater, it shouldn't need to know that.  Instead the work coordinator should be context aware - as in passed via construction, then `workItemsArePending` is called creates the appropriate context.

If we were passing a concrete context in function parameters, then I wouldn't have as much of an issue - but in this case a factory is being passed in.  While this is clever so a factory doesn't need to be provided into instance creation it makes this code very hard to interpret.

Ideally we would have code that follows a _single_ pattern of access to the context such as the following.  
```java
var docIndexer = new DocumentIndexer(context);
docIndexer.doIndexing();

// OR but less preferred

var docIndexer = new DocumentIndexer();
docIndexer.doIndexing(context);
```

File: RFS/src/main/java/com/rfs/cms/OpenSearchCmsClient.java
Comment (peternied) [inline RFS/src/main/java/com/rfs/cms/OpenSearchCmsClient.java:1]: Remove this empty file?

File: commonDependencyVersionConstraints/build.gradle
Comment (peternied) [inline commonDependencyVersionConstraints/build.gradle:104]: This should be clustered with the other `org.testcontainers` resources and reusing the variable
```suggestion
    api group: 'org.testcontainers', name: 'toxiproxy', version: testcontainers
```

File: RFS/src/main/java/com/rfs/tracing/RfsContexts.java
Comment (peternied) [reply RFS/src/main/java/com/rfs/tracing/RfsContexts.java]: Lets revisit this later, we've got many usages no need to stick on this instance here.

**Comment by peternied** [comment]
OK - did another full pass, there are still a number of unresolved comments, I think some of them are outdated, but you might want to double check.

I posted a number of IMO non-controversial minor changes, please take a quick pass at those.

For the comments that between me and approval are:
- Relationship between contexts: https://github.com/opensearch-project/opensearch-migrations/pull/678#discussion_r1672432432
- Responsibility of context creation: https://github.com/opensearch-project/opensearch-migrations/pull/678#discussion_r1672454102


image

File: DocumentsFromSnapshotMigration/src/main/java/com/rfs/RfsMigrateDocuments.java
Comment (peternied) [reply DocumentsFromSnapshotMigration/src/main/java/com/rfs/RfsMigrateDocuments.java:243]: [Offline Recap] - For this use case the WorkCoordinator has a slightly different use case than other parts of the system so lets run with this style.  Maybe we need to revisit in the future, but there will always be exceptions.  Marking this resolved.

File: DocumentsFromSnapshotMigration/src/main/java/com/rfs/RfsMigrateDocuments.java
Comment (peternied) [reply DocumentsFromSnapshotMigration/src/main/java/com/rfs/RfsMigrateDocuments.java:178]: [Offline Recap] - So yes this is generally true.  There is some slightly awkward terminology in the system, but there isn't any fundamental disagreement with how this works.  I'm fine moving forward with out any changes, resolving this thread.

File: DocumentsFromSnapshotMigration/src/test/java/com/rfs/ProcessLifecycleTest.java
Comment (gregschohn) [reply DocumentsFromSnapshotMigration/src/test/java/com/rfs/ProcessLifecycleTest.java:59]: Not with CsvSource.  I could have used methodSource, but that makes code less apparent.  I think that it's clearer to the reader to have the test cases just before the method.

File: DocumentsFromSnapshotMigration/src/main/java/com/rfs/RfsMigrateDocuments.java
Comment (gregschohn) [reply DocumentsFromSnapshotMigration/src/main/java/com/rfs/RfsMigrateDocuments.java:241]: The underlying difference is that the WorkCoordination code is its own independent subsystem from the rest of document migration.  Work coordination could be used from a number of different calling contexts, so the contexts that are relevant for each method of coordination may have different context ancestries.  That's the reason that a context factory instead is sent to each method rather than to the constructor.  The methods themselves could take a context instance as an arg, but then it would be up to the caller to close/autoclose the resource.

File: DocumentsFromSnapshotMigration/src/main/java/com/rfs/RfsMigrateDocuments.java
Comment (gregschohn) [reply DocumentsFromSnapshotMigration/src/main/java/com/rfs/RfsMigrateDocuments.java]: This is mostly true.  This part isn't quite...
> CoordinationContext is a child of the DocumentMigrationContext where it pulls configuration.

Both of those root contexts are 'root' to their callers.  Document migration contexts will be created from the root object or children that were returned by that root.  Wook coordination contexts will be created by the work coordination root context.   Those coordination contexts may have ancestor spans that are document migration spans.

For both 'root' contexts, they derive their configuration from the otel sdk, which, now in the latest code, is the same object - meaning that there's just one "service" in the otel-space for this process, which probably makes a lot more sense now.

**Review by peternied — APPROVED** [review]

**Pull request merged**
